File: algo_dart/binary_tree/binary_search_tree.dart
import 'binary_node.dart';

class BinarySearchTree<E extends Comparable<dynamic>> {
  BinaryNode<E>? root;

  void insert(E value) {
    root = _insertAt(root, value);
  }

  BinaryNode<E> _insertAt(BinaryNode<E>? node, E value) {
    if (node == null) {
      return BinaryNode(value: value);
    }
    if (value.compareTo(node.value) < 0) {
      node.leftChild = _insertAt(node.leftChild, value);
    } else {
      node.rightChild = _insertAt(node.rightChild, value);
    }

    return node;
  }

  @override
  String toString() {
    return root.toString();
  }
}

File: algo_dart/binary_tree/binary_search_tree.go
package binarytree

import (
	"fmt"

	"golang.org/x/exp/constraints"
)

type BinarySearchTree[E constraints.Ordered] struct {
	Root *BinaryNode[E]
}

func (bst *BinarySearchTree[E]) String() string {
	if bst.Root == nil {
		return "[]"
	}
	var result []E
	bst.Root.TraverseInOrder(func(value E) {
		result = append(result, value)
	})
	return fmt.Sprintf("%v", result)
}

func (bst *BinarySearchTree[E]) Insert(value E) {
	bst.Root = bst.insertAt(bst.Root, value)
}

func (bst *BinarySearchTree[E]) insertAt(node *BinaryNode[E], value E) *BinaryNode[E] {
	if node == nil {
		return &BinaryNode[E]{Value: value}
	}
	if value < node.Value {
		node.LeftChild = bst.insertAt(node.LeftChild, value)
	} else {
		node.RightChild = bst.insertAt(node.RightChild, value)
	}
	return node
}

func (bst *BinarySearchTree[E]) Contains(value E) bool {
	current := bst.Root

	for current != nil {
		if current.Value == value {
			return true
		}
		if value < current.Value {
			current = current.LeftChild
		} else {
			current = current.RightChild
		}
	}

	return false
}
